<a href="https://colab.research.google.com/github/bakut/packt-data-science-workshop/blob/master/Feature_Engineering_%E2%80%93_Exploration_of_Individual_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Feature Engineering – Exploration of Individual Features

explore the relationship between two variables, which are whether an individual owns a house and whether an individual has a loan, to the propensity for term deposit purchases by these individuals.

In [0]:
#importing pandas package
import pandas as pd

In [0]:
#assign link to a variable
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter03/bank-full.csv'

In [0]:
#read the banking dataset
bankData = pd.read_csv(file_url, sep=";")
bankData.head()

age           job  marital  education  ... pdays  previous poutcome   y
0   58    management  married   tertiary  ...    -1         0  unknown  no
1   44    technician   single  secondary  ...    -1         0  unknown  no
2   33  entrepreneur  married  secondary  ...    -1         0  unknown  no
3   47   blue-collar  married    unknown  ...    -1         0  unknown  no
4   33       unknown   single    unknown  ...    -1         0  unknown  no

[5 rows x 17 columns]

In [0]:
#relationship btw housing and propensity of term deposits
bankData.groupby(['housing', 'y'])['y'].agg(houseTot='count').reset_index()

housing    y  houseTot
0      no   no     16727
1      no  yes      3354
2     yes   no     23195
3     yes  yes      1935

In [0]:
# Relationship between having a loan and propensity for term deposits
bankData.groupby(['loan', 'y'])['y'].agg(loanTot='count').reset_index()

loan    y  loanTot
0   no   no    33162
1   no  yes     4805
2  yes   no     6760
3  yes  yes      484

In [0]:
#Taking the quantiles for 25%, 50% and 75% of the balance data
import numpy as np
np.quantile(bankData['balance'],[0.25,0.5,0.75])


array([  72.,  448., 1428.])

In [0]:
#converting the numerical values of the bank balance into categorical values
bankData['balanceClass'] = 'Quant1'
bankData.loc[(bankData['balance'] > 72) & (bankData['balance'] < 448), 'balanceClass'] = 'Quant2'
bankData.loc[(bankData['balance'] > 448) & (bankData['balance'] < 1428), 'balanceClass'] = 'Quant3'
bankData.loc[bankData['balance'] > 1428, 'balanceClass'] = 'Quant4'
bankData.head()

age           job  marital  education  ... previous  poutcome   y balanceClass
0   58    management  married   tertiary  ...        0   unknown  no       Quant4
1   44    technician   single  secondary  ...        0   unknown  no       Quant1
2   33  entrepreneur  married  secondary  ...        0   unknown  no       Quant1
3   47   blue-collar  married    unknown  ...        0   unknown  no       Quant4
4   33       unknown   single    unknown  ...        0   unknown  no       Quant1

[5 rows x 18 columns]

In [0]:
# Calculating the customers under each quantile
balanceTot = bankData.groupby(['balanceClass'])['y'].agg(balanceTot='count').reset_index()
balanceTot

balanceClass  balanceTot
0       Quant1       11340
1       Quant2       11275
2       Quant3       11299
3       Quant4       11297

In [0]:
# Calculating the total customers categorised as per quantile and propensity classification
balanceProp = bankData.groupby(['balanceClass', 'y'])['y'].agg(balanceCat='count').reset_index()
balanceProp

balanceClass    y  balanceCat
0       Quant1   no       10517
1       Quant1  yes         823
2       Quant2   no       10049
3       Quant2  yes        1226
4       Quant3   no        9884
5       Quant3  yes        1415
6       Quant4   no        9472
7       Quant4  yes        1825

In [0]:
# Merging both the data frames
balanceComb = pd.merge(balanceProp, balanceTot, on = ['balanceClass'])
balanceComb['catProp'] = (balanceComb.balanceCat / balanceComb.balanceTot)*100
balanceComb

balanceClass    y  balanceCat  balanceTot    catProp
0       Quant1   no       10517       11340  92.742504
1       Quant1  yes         823       11340   7.257496
2       Quant2   no       10049       11275  89.126386
3       Quant2  yes        1226       11275  10.873614
4       Quant3   no        9884       11299  87.476768
5       Quant3  yes        1415       11299  12.523232
6       Quant4   no        9472       11297  83.845269
7       Quant4  yes        1825       11297  16.154731

From the distribution of data, we can see that, as we move from Quantile 1 to Quantile 4, the proportion of customers who buy term deposits keeps on increasing. For instance, of all of the customers who belong to Quant 1, 7.25% have bought term deposits (we get this percentage from catProp). This proportion increases to 10.87 % for Quant 2 and thereafter to 12.52 % and 16.15% for Quant 3 and Quant4, respectively. From this trend, we can conclude that individuals with higher balances have more propensity for term deposits.

In this exercise, we explored the relationship of each variable to the propensity for term deposit purchases. The overall trend that we can observe is that people with more cash in hand (no loans and a higher balance) have a higher propensity to buy term deposits. In the next exercise, we will use these intuitions to derive a new feature.